### PR TITLE
Refactoring to use Identities instead of tax tags

### PR DESCRIPTION
**Pull request by samlown**
* Mexico now validating using Identity Keys instead of the previous Use tax tags.
* Identity objects will now be validated against the regime's list of acceptable "Identity Key Definitions", with the list of valid codes. This is somewhat basic validation at this time, just compares the list of codes, but could be expanded.

Codes now supported:
* `mx-cfdi-fiscal-regime` - Fiscal Regime Code, set in the Invoice Supplier *and* Customer
* `mx-cfdi-use` - Set as part of the customer's identities.
* `mx-cfdi-prod-serv` - Aligned with "ClaveProdServ" in the CFDI XML.

### CODE REVIEW

File: regimes/mx/identities.go [hunk 1]
```diff
@@ -0,0 +1,335 @@
+package mx
+
+import (
+	"github.com/invopop/gobl/i18n"
+	"github.com/invopop/gobl/tax"
+)
+
+// Regime specific keys for identities.
+const (
+	IdentityKeyFiscalRegime = "sat-fiscal-regime"
```
Comment (samlown) [inline regimes/mx/identities.go:10]: Consider renaming to `mx-fiscal-regime` to be more internationally compatible

File: regimes/mx/identities.go [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+	"github.com/invopop/gobl/i18n"
+	"github.com/invopop/gobl/tax"
+)
+
+// Mexican CFDI identity keys required by the SAT (tax authority in Mexico) in all
+// invoices and cannot be determined automatically.
+const (
+	IdentityKeyCFDIFiscalRegime = "mx-cfdi-fiscal-regime"
+	IdentityKeyCFDIUse          = "mx-cfdi-use"
+	IdentityKeyCFDIProdServ     = "mx-cfdi-prod-serv" // name from XML field: ClaveProdServ
```
Comment (samlown) [inline regimes/mx/identities.go:13]: I've gone with the `mx-cfdi-` prefix in the end. This makes it super clear what the context is, and namespaces the keys according to the regime.

**Comment by samlown** [comment]
I'm going to go ahead and merge this to create a new release. Please do comment regardless!

**Review by cavalle — COMMENTED** [review]
Epic PR! 🎩 A couple of quick thoughts, apart from the comments in the code, to be discussed and elaborated on over the next few days:

1. Principal concern: I think it would be more correct to give the UsoCFDI at invoice level rather than customer level. I think this because the same customer could have different values of UsoCFDI depending on the invoice. It wouldn't make sense for a supplier's stand-alone GOBL party document to specify a UsoCFDI in isolation. This suggests to me that the attribute belongs to the invoice, not the party.

2. A secondary concern is that I'm not sure it's conceptually correct (and therefore intuitive) to extend the concept of "identity" to classifications such as RegimenFiscal and UsoCFDI, which do not identify the entities (as [we still claim in the documentation](https://github.com/invopop/gobl/blob/main/org/party.go#L23)) but categorise them. The fact that we had to add a `key` to the `type` attribute, and implement separate validations/behaviours depending on which field is used, suggests to me that we are conflating two concepts that might be better separated. Either that, or if for economic reasons we don't need to add "just another struct with a key/type, a code, a label, a description…", maybe we should find a different and more generic word than "identity" to refer to this concept. I'll think about it some more and try to come up with some concrete suggestions.

**Comment by samlown** [comment]
@cavalle Thanks for looking at this in detail and for the comments! Here's my take:

1. UsoCFDI at the document level. This was the first implementation that aligned with the previous tax tags, but then while reading the CFDI specs, it became clear this is something that should be associated with the customer. While its true that in theory the value could change per invoice, this is very unlikely. A more likely scenario is that a single customer will have multiple very-similar invoices that should be issued under the same label. In my opinion, I think we should stick with it at the customer level reflecting that as a supplier, you're only likely to ask your customer once what their "Use" is, and every invoice after that initial decision will be the same. The alternative is asking for the "use" on each invoice, and that doesn't sound quite right.

2. "Identities" name isn't accurate enough. I'm inclined to agree. Also, in #184, to be able to achieve the same with the tax rates, a new name has been used: `codes` which is simply a map of a key to a code. It would be better to have a uniform implementation. We still have time to change this while we don't have many customers in MX.

I'll think about this in the background... something along the lines of `exts` referring to "extensions" or something similar could work. You want to infer that the data is not really part of GOBL, its something "extra" that is required to work locally. Pagero took this approach: https://pagero.github.io/puf-billing/#_country_specifics along the same lines.

**Comment by cavalle** [comment]
Good points! Regarding UsoCFDI, it's always hard for me to anticipate what would be more common. I imagine that the same customer, from the same supplier, buying the same type of product or service, will most of the time set the same UsoCFDI, yes. But I can see how the same customer is more likely to need to use a different UsoCFDI if the supplier (or the type of product/service they buy) varies. 

Also, one could argue that whether the same customer will use the same UseCFDI value or not should be a concern of the system providing a UI to the customer, not of the underlying struct that transports the data, which should be more concerned with modelling the domain in an intuitive way for developers. In this respect, a developer may be misled into thinking that the UsoCFDI is an attribute of the supplier, independent of the invoice (like the tax ID, the name or the RegimenFiscal are), when this isn't the case.

On the other hand, I have to admit that in the CFDI XML, the `UsoCFDI` attribute is part of the `Receptor` node, which does not support my point of view! 🤷

As for using extensions in GOBL, the same idea crossed my mind. That's a typical way to solve this problem in XML formats (e.g. UBL) which allow us to have structs more complex that key/values. The risk is that extensions could be abused and, as it happens with UBL and many others, everything ends up being an extension, and you end up with no standard. Also, you'll need additional JSON schemas, etc… But I'm not against the idea, in any case.

**Comment by samlown** [comment]
Hehe one option would be to support both. So effectively developers have the option of overriding the Customer's UsoCFDI at the invoice level, if they need to. I think what I worry about though, imagining the implementation here, is that if you don't have the "use" in the customer, you'll need to store it somewhere else. I'm not sure where that "somewhere else" is, unless you ask the user every time you build an invoice. In that scenario, it is just as easy to update the customer each time.

This morning I woke with the idea of a simple `ext` field that operates like a simple map. Kind of like the `meta`, but stricter. Examples in context:

```
{
  "$schema": "https://gobl.org/draft-0/bill/invoice",
  ....
  "tax": {
    "ext": {
      "mx-cfdi-use": "G01"
    }
  },
  ....
  "customer": {
    "tax_id": {
      "ext": {
        "mx-cfdi-fiscal-regime": "601"
      }
    }
  },
  .....
  "lines": [
    {
      "item": {
        "ext": {
          "mx-cfdi-prod-serv": "12345678"
        },
     },
     "tax": [
       {
          "cat": "VAT",
          "rate": "exempt",
          "ext": {
             "mx-cfdi-exempt": "XXX"
          },
      }
    ]
  }
}
```

I made up the `mx-cfdi-exempt` key, there are better examples for this in PT and IT. Basically, though, we're talking about a very simple extension key-value map, with validated keys and values. This would avoid as you say the complex extensions which are allowed in UBL/XML.

File: regimes/mx/item.go [hunk 3]
```diff
@@ -16,23 +16,28 @@ var (
...
 			}
 			return errors.New("SAT code must have 8 digits")
 		}
 	}
-
-	return errors.New("SAT code must be present")
+	return nil
 }
 
 func normalizeItem(item *org.Item) error {
```
[the remark this answers is not in the document]
Comment (cavalle) [reply regimes/mx/item.go:43]: Addressed in https://github.com/invopop/gobl/pull/186

File: regimes/mx/mx.go [hunk 4]
```diff
@@ -21,7 +21,6 @@ const (
 	KeySATFormaPago         cbc.Key = "sat-forma-pago"          // for mapping to c_FormaPago’s codes
 	KeySATTipoDeComprobante cbc.Key = "sat-tipo-de-comprobante" // for mapping to c_TipoDeComprobante’s codes
 	KeySATTipoRelacion      cbc.Key = "sat-tipo-relacion"       // for mapping to c_TipoRelacion’s codes
-	KeySATUsoCFDI           cbc.Key = "sat-uso-cfdi"            // for mapping to c_UsoCFDI’s codes
 
 	IdentityTypeSAT cbc.Code = "SAT" // for custom codes mapped from identities (e.g. c_ClaveProdServ’s codes)
```
[the remark this answers is not in the document]
Comment (cavalle) [reply regimes/mx/mx.go:25]: Addressed in https://github.com/invopop/gobl/pull/186

**Comment by cavalle** [comment]
> Hehe one option would be to support both

I've checked a couple of Mexican invoice apps (Facturama and Alegra) and they both have the "Uso CFDI" field as part of the client form (and not the invoice) (see for example [here](https://youtu.be/ZcIeXKqTeNQ)). So, as much as I still stand by my previous arguments, I'm now happy to leave this field as part of the supplier, and not at invoice level, if only to mirror the CFDI model and the way real-world clients seems to be implemented.

> Kind of like the meta, but stricter. 

Sounds like a good option to me! It may be accompanied by something in the regime data that allow us to define the valid keys, their labels and descriptions… similar to what you did with Identities. I wonder if these should completely replace org.Identities or not…

**Comment by samlown** [comment]
Thanks for investigating the Uso thing! So the customer seems okay for the time being, but we'll be mindful of feedback to see what could be useful.

> Sounds like a good option to me! It may be accompanied by something in the regime data that allow us to define the valid keys, their labels and descriptions… similar to what you did with Identities. I wonder if these should completely replace org.Identities or not…

I have a PR in progress to do precisely that. Its no longer using `org.Identities` for anything official. I'll upload as soon as I've had chance to iron out all the kinks.

**Pull request merged**